Dynamic config How-to update (#777)
rudderlabs/rudderstack-docs: docs/user-guides/how-to-guides/dynamic-destination-configuration.mdx

description: >-
  Configure your RudderStack destination dynamically via event payload properties.
---

Dynamic settings configuration is supported in <Link to="/destinations/rudderstack-connection-modes/#cloud-mode">cloud mode</Link> RudderStack destinations, allowing you to configure destinations based on data from an event payload.
## Create the configuration setting template

To use this feature, enter the path of your desired configuration value from the event payload using object notation in a [handlbars.js](https://handlebarsjs.com/guide/expressions.html) expression, as well as a default value for the setting in case the value is not present:

For example: `{{ event.path.to.key || "default_config_value" }}`, where

- `event` is the top level object for the event payload
- `path.to.key` is the location of the property you want to make configurable for your destination, e.g., `properties.url`
- `"default_config_value"` is the default value for the property

Any value present in the path will be passed as a configuration setting to your destination. If no value is present in the payload path, the default value will be used to configure the destination.

> If `event.path.to.key` exists, its value will be sent as a destination config setting. If not present, we will send the default value that you supply in the expression.

Reviewer: ```suggestion
If `event.path.to.key` exists, its value will be sent as a destination config setting. If not present, RudderStack will send the default value that you supply in the expression.
```

  </ul>
</div>

> In the next section we'll walk through an example of dynamic configuration using multiple Google Analytics tracking IDs.

Reviewer: ```suggestion
The next section walks through an example of dynamic configuration using multiple Google Analytics tracking IDs.
```

> ## Use Case: Dynamically configure your destination by Google Analytics Tracking ID

Reviewer 2: ```suggestion
## Use Case: Dynamically configure your Google Analytics destination using Tracking ID
```

> Suppose you have two web properties with `rudderanalytics.js` implemented, each with a unique Universal Analtyics (UA) tracking ID. You don't want to set up different RudderStack destinations for each property; rather, you would like to pass the event data to the appropriate tracking ID in your Google Analytics instance.

Reviewer 2: ```suggestion
Suppose you have implemented two web properties with `rudderanalytics.js`, each with a unique Universal Analytics (UA) tracking ID. You don't want to set up different RudderStack destinations for each property; rather, you would like to pass the event data to the appropriate tracking ID in your Google Analytics instance.
```

To achieve this you will:

1. Create a transformation that:
    - adds a `trackingId` property to your event
    - sets the appropriate `trackingId` value based on the RudderStack `sourceId` for the given web property.<br /><br />

1. Configure your Google Analytics destination in the RudderStack dashboard with a template that dynamically updates the `trackingId` property for each event, routing the data to the appropriate GA property.

### Step 1: Add a transformation

> To populate the configuration values from the event payload, add a <Link to="/features/transformations/#adding-a-transformation">transformation function</Link> to your Google Analytics destination.

Reviewer 2: ```suggestion
To populate the configuration values from the event payload, add a <Link to="/features/transformations/#adding-a-transformation">transformation function</Link>.
```

As we are adding Google Analytics destination in the next step.

The following sample code uses the transformation's `metadata()` function to determine the `sourceId` from the metadata object. If the `sourceId` matches one of your provided IDs, then the associated `trackingId` is sent to Google Analytics.

> If there is no match for the `sourceId`, the default value that you define in the next step will be passed to GA.

Reviewer: ```suggestion
If there is no match for the `sourceId`, the default value that you define in the next step will be passed to Google Analytics.
```

}
```

### Step 2: Specify the setting path in destination settings

> Once you have added a transformation, the next step is to <Link to="/dashboard-guides/destinations/#adding-a-destination">add a destination</Link> for Google Analytics in the RudderStack dashboard.

Reviewer 2: ```suggestion
The next step is to <Link to="/dashboard-guides/destinations/#adding-a-destination">add a Google Analytics destination</Link> and link the above-added transformation to it in the RudderStack dashboard.
```

> In the **Connection Settings**, the **Tracking ID** field of the destination should be configured as shown:

Reviewer: ```suggestion
In the **Connection Settings**, configure the **Tracking ID** field, as shown:
```

<img src="../../assets/user-guides/dynamic-destination-configuration.png" />

> With the transformation in place, and your tracking ID connection settings configured, event data will be routed to Google Analytics based on the tracking ID in the payload.

Reviewer: ```suggestion
With the transformation in place and the dynamic tracking ID configured in the connection settings, RudderStack will route the event data to Google Analytics based on the tracking ID in the payload.
```

## FAQ

> ### How is this feature useful to me?

Reviewer: @pete Do you think it would be useful to add this info at the start - highlighting how it allows you to add multiple instances of the destination programmatically, with different configs for specific fields?

Author: Yeah, i'll remove from FAQ and work in into the intro section 👍🏻

> This feature is useful if you want to configure multiple instances of the same destination with different configurations for specific fields. Refer to the use-case section above to see how you can use this feature.

Reviewer: ```suggestion
This feature is useful if you want to configure multiple instances of the same destination with different configurations for specific fields. Refer to the <Link to="/user-guides/how-to-guides/dynamic-destination-configuration/#use-case-dynamically-configure-your-destination-by-google-analytics-tracking-id">use case</Link> above to see how you can use this feature.
```

Reviewer 2: ```suggestion
This feature is useful if you want to configure multiple instances of the same destination with different configurations for specific fields. Refer to the <Link to="#use-case-dynamically-configure-your-destination-by-google-analytics-tracking-id">use case</Link> section above to see how you can use this feature.
```
